File: clients/ios/AwareChat-iOS/AwareChat-iOS-UnitTests/Core/Persistence/TestSupport/PersistenceTestStore.swift
// MARK: - AI Generated - Start
import Foundation
import SwiftData
@testable import AwareChat_iOS

@MainActor
final class PersistenceTestStore {
  let database: PersistenceContainer
  let users: SwiftDataUserLocalRepository
  let conversations: SwiftDataConversationLocalRepository
  let messages: SwiftDataMessageLocalRepository
  let writes: TestSaveControl

  init(url: URL? = nil) throws {
    let writes = TestSaveControl()
    self.writes = writes
    database = try PersistenceContainer(inMemory: url == nil, storeURL: url) { context in
      if writes.failNext {
        writes.failNext = false
        throw PersistenceError.writeFailed
      }
      try context.save()
      writes.successfulSaves += 1
    }
    users = SwiftDataUserLocalRepository(database: database)
    conversations = SwiftDataConversationLocalRepository(database: database)
    messages = SwiftDataMessageLocalRepository(database: database)
  }

  func identify() throws -> UserDTO {
    try users.saveIdentity(name: "Alice").wireIdentity()
  }

  func incoming(from peer: UUID = UUID(), id: UUID = UUID()) throws -> MessageDTO {
    let user = try identify()
    return try MessageDTO(
      messageId: id, text: "Hello", senderId: peer, receiverId: user.userId,
      clientCreatedAt: Date(timeIntervalSince1970: 100), clientSequence: 1,
      serverReceivedAt: Date(timeIntervalSince1970: 101)
    )
  }
}

@MainActor
final class TestSaveControl {
  var failNext = false
  var successfulSaves = 0
}
// MARK: - AI Generated - End

File: clients/android/AwareChat-Android/app/src/test/java/com/example/awarechat_android/core/persistence/testsupport/PersistenceTestStore.kt
package com.example.awarechat_android.core.persistence.testsupport

import android.app.Application
import com.example.awarechat_android.core.persistence.conversations.RoomConversationLocalRepository
import com.example.awarechat_android.core.persistence.database.AppDatabase
import com.example.awarechat_android.core.persistence.database.DatabaseFactory
import com.example.awarechat_android.core.persistence.database.PersistenceException
import com.example.awarechat_android.core.persistence.database.PersistenceFailure
import com.example.awarechat_android.core.persistence.messages.RoomMessageLocalRepository
import com.example.awarechat_android.core.persistence.users.RoomUserLocalRepository
import com.example.awarechat_android.core.protocol.MessageDto
import com.example.awarechat_android.core.protocol.UserDto
import java.time.Instant
import java.util.UUID
import java.util.concurrent.ConcurrentLinkedQueue
import org.junit.Assert.assertEquals
import org.junit.Assert.fail
import org.robolectric.RuntimeEnvironment

class PersistenceTestStore(
    databaseName: String? = null,
    identityIds: List<UUID> = listOf(DEFAULT_IDENTITY_ID),
    messageIds: List<UUID> = emptyList(),
) : AutoCloseable {
    private val identityIdQueue = ConcurrentLinkedQueue(identityIds)
    private val messageIdQueue = ConcurrentLinkedQueue(messageIds)
    val database: AppDatabase = if (databaseName == null) {
        DatabaseFactory.inMemory(application)
    } else {
        DatabaseFactory.create(application, databaseName)
    }
    val users = RoomUserLocalRepository(database) {
        identityIdQueue.poll() ?: UUID.randomUUID()
    }
    val conversations = RoomConversationLocalRepository(database)
    val messages = RoomMessageLocalRepository(database) {
        messageIdQueue.poll() ?: UUID.randomUUID()
    }

    suspend fun identify(name: String = "Alice"): UserDto = users.saveIdentity(name).wireIdentity()

    suspend fun incoming(
        peerId: UUID = UUID.randomUUID(),
        messageId: UUID = UUID.randomUUID(),
        sequence: Long = 1,
    ): MessageDto {
        val current = users.currentUser()?.wireIdentity() ?: identify()
        return MessageDto.create(
            messageId = messageId,
            text = "Hello",
            senderId = peerId,
            receiverId = current.userId,
            clientCreatedAt = Instant.ofEpochSecond(100),
            clientSequence = sequence,
            serverReceivedAt = Instant.ofEpochSecond(101),
        )
    }

    override fun close() {
        database.close()
    }

    companion object {
        val application: Application
            get() = RuntimeEnvironment.getApplication()

        val DEFAULT_IDENTITY_ID: UUID =
            UUID.fromString("11111111-1111-4111-8111-111111111111")
    }
}

suspend fun expectPersistenceFailure(
    expected: PersistenceFailure,
    operation: suspend () -> Unit,
) {
    try {
        operation()
        fail("Expected PersistenceException with $expected")
    } catch (error: PersistenceException) {
        assertEquals(expected, error.failure)
    }
}
